compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,340 @@ fn check_for_underconstrained_values_within_function(
+                result.extend(children);
+            }
+        }
+    }
+
+    /// If Brillig call is properly constrained by the given ids, return true
+    fn check_constrained(&self) -> bool {
+        // If every result has now been constrained,
+        // consider the call properly constrained
+        self.results.len() == self.results_constrained.len()
+    }
+
+    /// Remember partial constraints (involving some of the results and an argument)
+    /// along the way to take them into final consideration
+    fn store_partial_constraints(&mut self, constrained_values: &HashSet<ValueId>) {
+        let mut results_involved: Vec<usize> = vec![];
+
+        // For a valid partial constraint, a value descending from
+        // one of the results should be constrained
+        for (i, result_descendants) in self.results.iter().enumerate() {
+            // Skip checking already covered results
+            if self.results_constrained.contains(&i) {
+                continue;
+            }
+            if result_descendants.intersection(constrained_values).next().is_some() {
+                results_involved.push(i);
+            }
+        }
+
+        // Also, one of the argument descendants should be constrained
+        // (skipped if there were no arguments, or if an actual result and not a
+        // descendant has been constrained, e.g. against a constant)
+        if !results_involved.is_empty()
+            && (self.arguments.is_empty()
+                || self.root_results.intersection(constrained_values).next().is_some()
+                || self.arguments.intersection(constrained_values).next().is_some())
+        {
+            // Remember the partial constraint
+            self.results_constrained.extend(&results_involved);
+            // We can clear the unneeded sets now
+            results_involved.iter().for_each(|i| self.results[*i].clear());
+        }
+    }
+}
+
+impl DependencyContext {
+    /// Build the dependency context of variable ValueIds, storing
+    /// information on value ids involved in unchecked Brillig calls
+    fn build(&mut self, function: &Function, all_functions: &BTreeMap<FunctionId, Function>) {
+        self.block_queue.push(function.entry_block());
+        while let Some(block) = self.block_queue.pop() {
+            if self.visited_blocks.contains(&block) {
+                continue;
+            }
+            self.visited_blocks.insert(block);
+            self.process_instructions(block, function, all_functions);
+        }
+    }
+
+    /// Go over the given block tracking Brillig calls and checking them against
+    /// following constraints
+    fn process_instructions(
+        &mut self,
+        block: BasicBlockId,
+        function: &Function,
+        all_functions: &BTreeMap<FunctionId, Function>,
+    ) {
+        trace!("processing instructions of block {} of function {}", block, function);
+        let mut arguments = Vec::new();
+        let mut results = Vec::new();
```

Comment:
Any reason not to declare a fresh pair inside the loop?

Reply:
Extra allocations, although they probably get optimized out anyway. I don't think it matters a lot actually though. You thinking bringing them into the loop for clarity?

Reply:
Yeah, just because then we don't have to think about how these are used after the loop, which they aren't. I don't typically see loop variables bubbling up to outside the loop very often.

Reply:
Okay